File: Heap/Python/Maximum Average Pass Ratio.py
# Probem link: https://leetcode.com/problems/maximum-average-pass-ratio/description/
# Time: O(n log n)
# Space: O(n)
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        store = []

        for ps, stu in classes:
            avg = ps / stu
            inc = (ps + 1) / (stu + 1) - avg
            store.append((-inc, ps, stu))
        
        heapify(store)

        for _ in range(extraStudents):
            _, ps, stu = heappop(store)
            avg = (ps + 1) / (stu + 1)
            newInc = (ps + 2) / (stu + 2) - avg
            heappush(store, (-newInc, ps + 1, stu + 1))
        
        return sum(ps / stu for _, ps, stu in store) / len(classes)

File: Heap/Java/Maximum Average Pass Ratio.java
// Problem link: https://leetcode.com/problems/maximum-average-pass-ratio/description/
// Time: O(n log n)
// Space: O(n)
class Solution {
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        PriorityQueue<double[]> store = new PriorityQueue<>((a, b) -> Double.compare(b[0], a[0]));
        double avg, inc, result = 0, pop[], pass, total;

        for(int[] cls : classes) {
            pass = cls[0];
            total = cls[1];
            avg = pass / total;
            inc = (pass + 1) / (total + 1) - avg;
            store.offer(new double[] {inc, pass, total});
        }

        for(int student = 0; student < extraStudents; student++) {
            pop = store.poll();
            avg = (pop[1] + 1) / (pop[2] + 1);
            inc = (pop[1] + 2) / (pop[2] + 2) - avg;
            store.offer(new double[] {inc, pop[1] + 1, pop[2] + 1});
        }

        while(!store.isEmpty()) {
            pop = store.poll();
            result += pop[1] / pop[2];
        }

        return result / classes.length;
    }
}
